Mage.Sets/src/mage/cards/u/UnboundFlourishing.java

```diff
@@ -50,38 +47,73 @@ public UnboundFlourishing copy() {
-        staticText = "Whenever you cast a permanent spell with a mana cost that contains {X}, double the value of X";
+    UnboundFlourishingDoubleXAbility() {
+        super(Zone.BATTLEFIELD, new UnboundFlourishingDoubleXEffect(), false);
+        setTriggerPhrase("Whenever you cast a permanent spell with a mana cost that contains {X}");
     }
 
-    private UnboundFlourishingDoubleXEffect(final UnboundFlourishingDoubleXEffect effect) {
-        super(effect);
+    private UnboundFlourishingDoubleXAbility(final UnboundFlourishingDoubleXAbility ability) {
+        super(ability);
     }
 
     @Override
-    public boolean replaceEvent(GameEvent event, Ability source, Game game) {
-        event.setAmount(CardUtil.overflowMultiply(event.getAmount(), 2));
-        return false;
+    public UnboundFlourishingDoubleXAbility copy() {
+        return new UnboundFlourishingDoubleXAbility(this);
     }
 
     @Override
-    public boolean checksEventType(GameEvent event, Game game) {
-        return event.getType() == GameEvent.EventType.X_MANA_ANNOUNCE;
+    public boolean checkEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.SPELL_CAST;
     }
 
     @Override
-    public boolean applies(GameEvent event, Ability source, Game game) {
-        Spell spell = game.getSpell(event.getTargetId());
-        return spell != null && spell.isPermanent(game) && spell.isControlledBy(source.getControllerId());
+    public boolean checkTrigger(GameEvent event, Game game) {
+        if (!event.getPlayerId().equals(getControllerId())) {
+            return false;
+        }
+        
+        Spell spell = game.getStack().getSpell(event.getTargetId());
+        if (spell != null && spell.isPermanent(game)) {
+            if (spell.getSpellAbility().getManaCostsToPay().containsX()) {
+                game.getState().setValue(this.getSourceId() + UnboundFlourishing.needPrefix, spell);
+                return true;
+            }
+        }
+        return false;
+    }
+}
+
+class UnboundFlourishingDoubleXEffect extends OneShotEffect {
+
+    UnboundFlourishingDoubleXEffect() {
+        super(Outcome.Benefit);
+        this.staticText = ", double the value of X";
+    }
+
+    private UnboundFlourishingDoubleXEffect(final UnboundFlourishingDoubleXEffect effect) {
+        super(effect);
     }
 
     @Override
     public UnboundFlourishingDoubleXEffect copy() {
         return new UnboundFlourishingDoubleXEffect(this);
     }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        Player controller = game.getPlayer(source.getControllerId());
+        if (player != null && controller != null) {
+            Spell needObject = (Spell) game.getState().getValue(source.getSourceId() + UnboundFlourishing.needPrefix);
+            Map<String, Object> tagsMap = CardUtil.getSourceCostsTagsMap(game, needObject.getSpellAbility());
```

Comment:
Don't change saved objects. It's can be potentially bugged (example: saved object contains own list of cost tags and can be changed from different places). It's better to use real objects -- e.g. save spell id and search it in stack for doubling.